[CALCITE-1703] Functions on TIMESTAMP column throws ClassCastException

ijokarumawak opened the pull request:
Currently, TIMESTAMP or TIME column is used at any function, ClassCastException is thrown. For example, if a TIMESTAMP column `date_added` is used like this, `select name, EXTRACT(YEAR FROM date_added) as year_added from FLOWFILE`, Calcite generates the following code. And a ClassCastException is thrown at `(Long) current[1]`:

```Java
            public Object current() {
              final Object[] current = (Object[]) inputEnumerator.current();
              return new Object[] {
                  current[0],
                  (java.sql.Timestamp) current[1] == null ? (Long) null : Long.valueOf(org.apache.calcite.avatica.util.DateTimeUtils.unixDateExtract(org.apache.calcite.avatica.util.TimeUnitRange.YEAR, (Long) current[1] / 86400000L))};
            }
```

That happens because of the wrong if condition at `PhysTypeImple.fieldReference` method where it ignores the fieldType that is used as a fromType to convert the value to Long. Since fromType is null, the generated code casts `current[1]` to Long blindly.

The condition should accept not only `java.sql.Date` but also `java.sql.Time` and `java.sql.Timestamp`. Since these classes are all sub-class of `java.util.Date`, using `isAssignableFrom` method will address the issue.

Current if statement only supports `java.sql.Date`:
![image](https://user-images.githubusercontent.com/1107620/49925417-363d4980-fefc-11e8-8675-78c68b540da7.png)

DISCLAIMER: This is my first PR to Apache Calcite project. Kindly give me some advice on unit testing.

I wanted to add an unit test to reproduce the issue and prove this fix it. But I couldn't figure out where is the right place for such test.

There is an existing case [testExtractYearFromTimestamp ](https://github.com/apache/calcite/blob/master/core/src/test/java/org/apache/calcite/test/JdbcTest.java#L1474) but this test doesn't call `PhysTypeImple.fieldReference`. Some other test cases in the same JdbcTest such as `testSumOverUnboundedPreceding` does call `PhysTypeImple.fieldReference`. I don't know where this difference comes from..

ijokarumawak committed:
[CALCITE-1703] Functions on TIME or TIMESTAMP column can throw ClassCastException

Incorporated review comments

- Added the new test to CalciteSuite
- Added test case message to clarify assertion objectives

Adding more fixes

- java.sql.Time is not converted to int correctly
- java.sql.Timestamp fraction can not be used with some conditions such as '='

Fixed checkstyle errors.

Better assertion messages.

ijokarumawak commented:
I found that the particular code branch that can throw ClassCastException is generated when a timestamp column in an Object array row type is referenced. Referencing timestamp field at a Java class (POJO) doesn't seem having this issue.

ijokarumawak commented:
Fixed check-style errors.

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -0,0 +1,115 @@
+
+      ResultSet resultSet = statement.executeQuery(sql);
+      assertTrue(resultSet.next());
+      assertEquals(2018, resultSet.getInt(1));
```
Would you please add a message to `assertEquals`?
Assert failure should look like a decent bug report, and current `assertEquals(2018, resultSet.getInt(1));` would produce something like `expected 2018 got 0`.

The other way to look at "message" parameter: would you be able to tell the nature of the bug by looking at error message+stacktrace only?

ijokarumawak replied:
Makes sense. I've added messages. Thanks!

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Test for querying Java Object array table.
+ */
+public class ObjectArrayTableTest {
```
Would you please add the test to a test suite?

ijokarumawak replied:
Added it to CalciteSuite. The test took 1.4 sec on my Macbook.

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -0,0 +1,115 @@
+      assertEquals(2018, resultSet.getInt(1));
+      // Although using 'extract' function against time value doesn't make sense,
+      // this confirms time column can be converted to int.
+      assertEquals(0, resultSet.getInt(2));
```
Is this consistent across various time zones?

ijokarumawak replied:
I've tested it with different timezones and the results were consistent. I believe the test is timezone independent as those date-times are computed with local timezone.

ijokarumawak commented:
@vlsi Thanks for reviewing! I've added more commits to address your comments. As I was working on that, I've found few more issues to fix. The added test cases illustrate what had not been working as expected.

ijokarumawak commented on core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java:
```diff
@@ -651,6 +651,14 @@ public SqlTypeName getSqlTypeName() {
+        return 9;
+      }
+      return super.getPrecision();
+    }
```
Without this, Timestamp fraction part gets lost when the value is cast to TIMESTAMP (without precision). This affects where clause conditions with '='. E.g. `where \"ts\" = TIMESTAMP '2018-12-14 18:29:34.123'` does not match if the 'ts' has the same timestamp because the fraction part `123` dropped.

vlsi replied:
I wonder what should be the result of `cast(timestamp '2018-12-14 18:29:34.123' as timestamp(2)) = timestamp '2018-12-14 18:29:34.123'`

@ijokarumawak , have you though of that case?

By the way, I appreciate your comment here, however it would be so much better to have that comment as a Java comment in the sources.

ijokarumawak replied:
Good point. I haven't thought casting with a precision..
Now I'm adding unit test cases for that. Ideally casting java.sql.Timestamp should work the same as timestamp literals.

I'm going to add the comment to the source, too.

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -0,0 +1,239 @@
+
+      ResultSet resultSet = statement.executeQuery(sql);
+      assertTrue(resultSet.next());
+      assertEquals("java.sql.Date is converted to int when it's passed to a function",
```
Thank you for adding the message, however:
1) "conversion to int" is a current implementation detail, so it might easily change
2) The test does NOT verify if `Date` is converted to int or not. The test verifies that end-to-end value is valid (it is good by the way!)

Current test would fail like

```
AssertionError: java.sql.Date is converted to int when it's passed to a function, got: 2017, expected: 2018
```

I don't think the message clarifies why 2018 was expected.

I suggest message should be like

`assertEquals("extract(year from Date.valueOf(\"2018-12-14\")", 2018, resultSet.getInt(1));`

In case of failure, it will fail like the following:

```
AssertionError: extract(year from Date.valueOf("2018-12-14"), got: 2017, expected: 2018
```

That enables to manually check if the expected value is indeed right.

ijokarumawak replied:
Thanks for clarifying the idea. That makes sense. I've updated the assertion messages.

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -0,0 +1,239 @@
+
+      Statement statement = connection.createStatement();
+      ResultSet resultSet = statement.executeQuery(sql + " where \"tm\" > TIME '18:29:33'");
+      assertTrue("Time is grater than the literal", resultSet.next());
```
The failure of this assertion would be quite cryptic.
Something like `assertTrue("sql.Time.valueOf("18:29:34") > TIME '18:29:33'", ...)` would be better.

ijokarumawak replied:
Done. Thanks!

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -0,0 +1,239 @@
+
+      statement = connection.createStatement();
+      resultSet = statement.executeQuery(sql + " where \"tm\" = TIME '18:29:34'");
+      assertTrue("Time equals to the literal", resultSet.next());
```
Have you considered refactoring "createStatement + executeQuery + assert" to a separate method?
It looks like `"\"tm\" = TIME '18:29:34'"` could be method parameter so it is used in both `assertMessage` and query text.

ijokarumawak replied:
Consider that a bit, but I just updated the assertion messages. Thanks for the suggestion anyway.

ijokarumawak commented:
@F21 Yes, I believe this PR covers what #405 tries to fix.

ijokarumawak commented:
Rebased with the latest master as the PR had a conflict on CalciteSuite. Also squashed commits.

ijokarumawak committed:
Fixed time fraction cast.

ijokarumawak commented:
@vlsi I've implemented timestamp cast and added unit test for that. The issue was the existing code has had the conversion but used `scale` instead of `precision`. Please see the added commit for what I changed. Hopefully this PR is now ready to be merged. Thanks!

risdenk commented:
@vlsi - planning to merge still?

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -241,6 +244,35 @@ private void withJavaSqlDateTypes(TestWithConnection test) throws SQLException {
+            format(Locale.ROOT, sql, expected, actual));
+        resultSet.next();
+        assertEquals(format(Locale.ROOT, "cast(java.sql.Timestamp as %s)", castTo),
+            resultSet.getTimestamp(1), resultSet.getTimestamp(2));
```
@ijokarumawak , this is wrong. "EXPECTED" value must not come from a system under test.
You should either hard-code the expected value (which is preferred) or compute it with a "simple to understand" approach.

ijokarumawak replied:
Changed to use hard-coded expected values.

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -241,6 +244,35 @@ private void withJavaSqlDateTypes(TestWithConnection test) throws SQLException {
+        final ResultSet resultSet = statement.executeQuery(
+            format(Locale.ROOT, sql, expected, actual));
+        resultSet.next();
+        assertEquals(format(Locale.ROOT, "cast(java.sql.Timestamp as %s)", castTo),
```
It is great you add `castTo` to the message, however it would be really good if you add source value as well.

ijokarumawak replied:
Added source value.

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -241,6 +244,35 @@ private void withJavaSqlDateTypes(TestWithConnection test) throws SQLException {
+      final String sql = "select %s as EXPECTED, %s as ACTUAL from \"java_sql_date_types\"";
+
+      for (int i = 3; i >= 0; i--) {
+        final String castTo = format(Locale.ROOT, "TIMESTAMP(%d)", i);
```
You seem to use `format(Locale.ROOT` extensively. It makes the code extremely obscure. Could you just use `+` operator?

Alternative option is to use Kotlin's string interpolation

ijokarumawak replied:
Changed to simple string concatenation using `+`.

vlsi commented on core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java:
```diff
@@ -241,6 +244,35 @@ private void withJavaSqlDateTypes(TestWithConnection test) throws SQLException {
+  @Test public void testCastTimestampToTimestamp() throws SQLException {
+
+    withJavaSqlDateTypes(connection -> {
+
```
Please avoid having to many blank lines

ijokarumawak replied:
Removed blank lines.

ijokarumawak committed:
Incorporated review comments

ijokarumawak commented:
Hi @vlsi, sorry for the delay. I've added another commit to address review feedback. Please let me know if there's anything more. Thanks for reviewing!

ijokarumawak commented:
Hi @vlsi I just noticed that CI tests failed at `DateRangeRulesTest`. But the only thing I changed at the last commit is `ObjectArrayTableTest` test class. Do you know why `DateRangeRulesTest` failed on travis and appveyor?? The test had been successful before, and is successful if I run it locally..

michaelmior commented:
@vlsi Could you take another look?

vlsi commented:
@michaelmior , the review comments are resolved.

However it looks like the change fails the build, and I have not analyzed the reason.

michaelmior commented:
Thanks!

michaelmior commented:
@ijokarumawak What JDK version are you using? It passes on CI with JDK 10.

vlsi commented:
@michaelmior , it does not.
I think the reason one of Travis job passes is it runs with `SLOW_TESTS=Y`, so only a subset of tests is executed.
